ACM#9987: HCP: Document Agent provisioning flow for non BM machines (#6073)
open-cluster-management/rhacm-docs: clusters/hosted_control_planes/handle_ingress_non_bm.adoc

[#handling-ingress-non-bm]
= Handling ingress in a hosted cluster on non bare metal agent machines

> Every {ocp-short} cluster comes set up with a default application Ingress Controller that is expected have an external DNS record associated with it. For example, if you create a HyperShift cluster named `example` with the base domain `krnl.es`, you can expect the wildcard domain

Reviewer: ```suggestion
Every {ocp-short} cluster has a default application Ingress Controller that is expected to have an external DNS record associated with it. For example, if you create a HyperShift cluster named `example` with the base domain `krnl.es`, you can expect the wildcard domain
```

`*.apps.example.krnl.es` to be routable.

> You can set up a load balancer and wildcard DNS record for the `*.apps`. This process requires deploying MetalLB, configuring a new load balancer service that routes to the ingress deployment, and assigning a wildcard DNS entry to the load balancer IP address.

Reviewer: ```suggestion
You can set up a load balancer and wildcard DNS record for the `*.apps`. The process requires deploying MetalLB, configuring a new load balancer service that routes to the ingress deployment, and assigning a wildcard DNS entry to the load balancer IP address. Complete the following steps:
```

.. Create a YAML file that contains the configuration for the MetalLB instance:

+
[source,yaml]
----
apiVersion: metallb.io/v1beta1
kind: MetalLB
metadata:
  name: metallb
  namespace: metallb
----

.. Save the file as `metallb-instance-config.yaml`.

.. Create the MetalLB instance by entering this command:

+
----
oc apply -f metallb-instance-config.yaml
----

. Configure the MetalLB Operator by creating two resources:

+
** An `IPAddressPool` resource with a single IP address. This IP address must be on the same subnet as the network that the cluster nodes use.
** A `BGPAdvertisement` resource to advertise the load balancer IP addresses that the `IPAddressPool` resource provides through the BGP protocol.
+
**Important:** Change the `INGRESS_IP` environment variable to match your environment's address.

+
.. Create a YAML file to contain the configuration:

+
[source,yaml]
----
export INGRESS_IP=192.168.122.23

apiVersion: metallb.io/v1beta1
kind: IPAddressPool
metadata:
  name: ingress-public-ip
  namespace: metallb
spec:
  protocol: layer2
  autoAssign: false
  addresses:
    - ${INGRESS_IP}-${INGRESS_IP}
---
apiVersion: metallb.io/v1beta1

> kind: BGPAdvertisement

Reviewer 2: The BGPAdvertisement works for you? In my env I've only make the L2Advertisement work properly.

Author: sry, I haven't tried `BGPAdvertisement`. This module is copied from the [pre-existing bare metal section](https://github.com/stolostron/rhacm-docs/blob/2.10_stage/clusters/hosted_control_planes/handle_ingress_bm.adoc)

+
[source,yaml]
----
kind: Service
apiVersion: v1
metadata:
  annotations:
    metallb.universe.tf/address-pool: ingress-public-ip
  name: metallb-ingress
  namespace: openshift-ingress
spec:
  ports:
    - name: http
      protocol: TCP
      port: 80
      targetPort: 80
    - name: https
      protocol: TCP
      port: 443
      targetPort: 443
  selector:
    ingresscontroller.operator.openshift.io/deployment-ingresscontroller: default
  type: LoadBalancer
----

.. Save the file as `metallb-loadbalancer-service.yaml`.

.. Enter the following command to apply the configuration from the YAML file:

+
----
oc apply -f metallb-loadbalancer-service.yaml
----

.. Enter the following command to reach the {ocp-short} console:

+
----
curl -kI https://console-openshift-console.apps.example.krnl.es

HTTP/1.1 200 OK
----

.. Check the `clusterversion` and `clusteroperator` values to verify that everything is running. Enter the following command:

+
----
oc --kubeconfig ${HOSTED_CLUSTER_NAME}.kubeconfig get clusterversion,co
----

+
See the following example output:

+
----
NAME                                         VERSION   AVAILABLE   PROGRESSING   SINCE   STATUS

> clusterversion.config.openshift.io/version   4.12z    True        False         3m32s   Cluster version is 4.12z

Reviewer 2: I would change the 4.12z to the version you mentioned at the beginning of the PR. Have in mind to do that also in the nexts ones.

Author: Updated `4.12.z` to `4.14.13-x86_64`

Reviewer 3: Same here, we are using variables when we can.

clusteroperator.config.openshift.io/openshift-apiserver                        4.14.13-x86_64    True        False         False      54m
clusteroperator.config.openshift.io/openshift-controller-manager               4.14.13-x86_64    True        False         False      54m
clusteroperator.config.openshift.io/openshift-samples                          4.14.13-x86_64    True        False         False      23m
clusteroperator.config.openshift.io/operator-lifecycle-manager                 4.14.13-x86_64    True        False         False      54m
clusteroperator.config.openshift.io/operator-lifecycle-manager-catalog         4.14.13-x86_64    True        False         False      54m
clusteroperator.config.openshift.io/operator-lifecycle-manager-packageserver   4.14.13-x86_64    True        False         False      54m
clusteroperator.config.openshift.io/service-ca                                 4.14.13-x86_64    True        False         False      25m
clusteroperator.config.openshift.io/storage                                    4.14.13-x86_64    True        False         False      25m

> ----

Reviewer: Not sure what we are looking for here but we need to shorten/remove this output as well and let the user know what they need to look for.
